src/vs/editor/common/viewLayout/viewLineRenderer.ts

```diff
@@ -540,6 +562,11 @@ function _applyRenderWhitespace(lineContent: string, len: number, continuesWithW
 			isInWhitespace = false;
 		}
 
+		// If rendering whitespace on selection, check that the charIndex falls within a selection
+		if (isInWhitespace && selections) {
+			isInWhitespace = selections.some(selection => selection.containsPosition(new Position(selection.startLineNumber, charIndex + 1)));
```

Comment:
The line rendering code is very hot and I would like to keep it as performant as possible.

Could we please eliminate this `O(LN)` where `N` is the selection length and `L` is the line character length? For each character, the `selections` array is potentially walked to the end to check if it contains the position. 

I suggest you change to something like the following to avoid this inner loop:
- ensure the selection pieces are sorted. That can be done outside this method, I would do it in `RenderLineInput`'s ctor to make sure that `equals` has a chance to save some rendering even if the selection pieces come in unsorted.
- in this method, have an outer variable, `selectionPieceIndex` that tracks the current selection piece index. This index can be incremented whenever leaving the current selection piece.
- then a single `if` is sufficient to check if the current character is inside the current selection piece.